Framework/Geometry/src/Instrument/ParameterMap.cpp

```diff
@@ -1199,20 +1227,30 @@ size_t ParameterMap::componentIndex(const ComponentID componentId) const {
   return m_componentInfo->indexOf(componentId);
 }
 
-/// Only for use by ExperimentInfo. Sets the pointer to the DetectorInfo.
-void ParameterMap::setDetectorInfo(
-    boost::shared_ptr<const Beamline::DetectorInfo> detectorInfo) {
-  m_detectorInfo = std::move(detectorInfo);
-}
-
-void ParameterMap::setComponentInfo(
-    boost::shared_ptr<const Geometry::ComponentInfo> componentInfo) {
-  m_componentInfo = std::move(componentInfo);
-}
-
 /// Only for use by Instrument. Sets the pointer to the owning instrument.
 void ParameterMap::setInstrument(const Instrument *instrument) {
+  if (instrument == m_instrument)
+    return;
+  if (!instrument) {
+    m_componentInfo = nullptr;
+    m_detectorInfo = nullptr;
+    return;
+  }
+  if (m_instrument)
+    throw std::logic_error("ParameterMap::setInstrument: Cannot change "
+                           "instrument once it has been set.");
+  if (instrument->isParametrized())
+    throw std::logic_error("ParameterMap::setInstrument must be called with "
+                           "base instrument, not a parametrized instrument");
   m_instrument = instrument;
+
+  const auto parInstrument = ParComponentFactory::createInstrument(
+      boost::shared_ptr<const Instrument>(m_instrument, NoDeleting()),
+      boost::shared_ptr<ParameterMap>(this, NoDeleting()));
+
+  InstrumentVisitor visitor(parInstrument);
```

Comment:
I tried now, combining it with the even less nice code in the copy constructor, but that did not improve things (mainly setting up the cross links is messy). I'd suggest to leave this until we figure out a cleaner solution for copying the two Info classes safely while maintaing correct links between them, and to the instrument.